include/ibamr/WaveGenerationStrategy.h

```diff
@@ -0,0 +1,233 @@
+     */
+    SAMRAI::tbox::Pointer<SAMRAI::hier::VariableContext> d_phi_new_ctx;
+
+    /*
+     * Start and end coordinates of the damping zone, and the damping coefficient.
+     */
+    double d_x_zone_start, d_x_zone_end;
+    double d_alpha;
+    int d_sign_gas_phase = 1;
+
+    /*!
+     * Number of interface cells to represent air-water thickness.
+     */
+    double d_num_interface_cells;
+};
+
+/*!
+ * Base class for generating Stokes wave.
+ */
+class StokesWaveGenerator
+{
+public:
+    /*!
+     * \brief Constructor of the class.
+     */
+    StokesWaveGenerator(const std::string& object_name, SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> input_db);
+
+    /*!
+     * \brief Destructor of the class.
+     */
+    virtual ~StokesWaveGenerator() = default;
+
+    /*!
+     * Get surface elevation at a specified horizontal position and time.
+     */
+    virtual double getSurfaceElevation(double x, double time) const = 0;
+
+    /*!
+     * Get velocity component at a specified position and time.
+     */
+    virtual double getVelocity(double x, double z_plus_d, double time, int comp_idx) const = 0;
+
+    /*
+     * \brief Get Stokes wave parameters.
+     */
+    double getWaterDepth() const
+    {
+        return d_depth;
+    } // getWaterDept
+
+    double getWaveAngularFrequency() const
+    {
+        return d_omega;
+    } // getWaveAngularFrequency
+
+    double getWaveNumber() const
+    {
+        return d_wave_number;
+    } // getWaveNumber
+
+    double getWaveAmplitude() const
+    {
+        return d_amplitude;
+    } // getWaveAmplitude
+
+    double getGravity() const
+    {
+        return d_gravity;
+    } // getGravity
+
+    /*
+     * \brief WaveGeneration data object.
+     */
+    WaveGenerationData d_wave_gen_data;
+
+protected:
+
+
+    /*!
+     * Book-keeping.
+     */
+    std::string d_object_name;
+
+    /*!
+     * \brief Wave parameters.
+     *
+     * \param d_wave_number  : Wave number of dominant wave component [$2\pi/m$]
+     * \param d_amplitude    : Amplitude of the dominant wave component [m]
+     * \param d_depth        : Depth of water, from sea bed to still water level [m]
+     * \param d_gravity      : Acceleration due to gravity [$m/s^2$]
+     * \param d_omega        : Angular frequency [$2 \pi/s$] (optional)
+     *
+     * \NOTE The class calculates a more accurate value of omega from the expansion coefficients
+     * and the provided value in not used.
+     */
+    double d_depth, d_omega, d_wave_number, d_amplitude, d_gravity;
+
+    /*!
+     * If we are calculating in deep water limit.
+     */
+    bool d_deep_water_limit = false;
+
+    /*
+     *  \brief Stokes wave generator data members.
+     */
+    /*!
+     * Get wave parameters from input db.
+     */
+    void getFromInput(SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> db);
+
+}; // StokesWaveGenerator
+
+class FifthOrderStokesWaveGenerator : public StokesWaveGenerator
```

Comment:
Usually we would have each of these classes in their own headers, so that users do not have to guess which header provides which classes. Can we split out the Stokes wave generator?

Reply:
I think we put all Stokes wave classes (1st, 2nd and 5th) in a single header. Other kinds of waves will be in different header. Does that sound good?

Reply:
I think it is better to split them out into individual headers. But, sounds OK.

Reply:
Upon further reflection ... I would prefer to keep with convention that classes that are intended to be used by users go into their own headers.